File: infinitum/euler.c
#include<stdio.h>
#include<math.h>
long long int power(int a, int b)
{
	long long int c=1,temp=1000000000000000,m=(2*b)+1;
	while(b!=0)
	{
		c=c*a;
	/*	if((c/temp)!=0)
		{
			
			while((c/temp)!=0)
			{
				long long int p=c/temp;
				c = c%temp;
				p = p%m;
				p = p*temp;
				c = c+p;
			}
		}*/
		b--;
	}
	printf("c=%lld\n",c);
	return c;
}
int main()
{
	int t,q;
	scanf("%d",&t);
	for(q=0;q<t;q++)
	{
		int a,m;
		long long int ans;
		scanf("%d %d",&a,&m);
		ans=power(a,(m-1)/2);
			printf("ans=%lld\n",ans);
//		if(ans/temp==0)
			ans=(ans-1)%m;
			if(ans==0)
				printf("YES\n");
			else
				printf("NO\n");
	/*	else
		{
			while(ans/temp==0)
			{
				long long int p=ans/temp;
				p = p%m;
				p = p*temp;
				ans = ans+p;
			}
			ans=(ans-1)%m;
			if(ans==0)
				printf("YES\n");
			else
				printf("NO\n");
		}
*/
	}
	return 0;
}

File: infinitum/euler.py
def power(b,e):
	r=1
	while e:
		if(e&1):
			r *= b
		e >>=1
		b *= b
	return r

def is_prime(n):
	if n == 2 or n == 3: return True
	if n < 2 or n%2 == 0: return False
	if n < 9: return True
	if n%3 == 0: return False
	r = int(n**0.5)
	f = 5
	while f <= r:
#print '\t',f
		if n%f == 0: return False
		if n%(f+2) == 0: return False
		f +=6
	return True

t = raw_input()
t = int(t)
for q in xrange(0,t):
	a,m = raw_input().split()
	a = int(a)
	m = int(m)
	if(a>m):
		print "NO"
	elif(is_prime(m)==False):
		print "NO"
	else:
		ans = power(a,(m-1)/2)
		ans = (1-ans)%m
		if(ans==0):
			print "YES"
		
		else:
			print "NO"

